### ISSUE TITLE
Allow Multiple Legends to be Displayed

### ISSUE THREAD

**Issue by CIOData** [walkerke/mapgl#45]
I have an application where I desire to have a legend displaying values from a choropleth map and another legend displaying values for a circle layer on top of the choropleth map. However, only the last legend generated displays. I tried using `unique_id` to have them display as different elements in the CSS, but this didn't make a difference.

**Comment by walkerke** [comment]
I definitely see this use-case, should be a quick fix!

**Comment by walkerke** [comment]
Try out the new argument `add = TRUE` when you add a legend:

```r
library(mapgl)

map <- mapboxgl(
  center = c(-96, 37.8),
  zoom = 3
)

# Add a categorical legend for population
map <- map |>
  add_legend(
    legend_title = "Population",
    values = c("Low", "Medium", "High"),
    colors = c("#FED976", "#FEB24C", "#FD8D3C"),
    type = "categorical",
    circular_patches = TRUE,
    position = "top-left",
    sizes = c(4, 8, 12)
  )

# Add another categorical legend for land use
map <- map |>
  add_legend(
    legend_title = "Land Use",
    values = c("Urban", "Agricultural", "Forest", "Water"),
    colors = c("#E31A1C", "#33A02C", "#1F78B4", "#A6CEE3"),
    type = "categorical",
    circular_patches = FALSE,
    position = "top-right", 
    add = TRUE
  )

# Add a continuous legend for temperature
map <- map |>
  add_legend(
    legend_title = "Temperature (°C)",
    values = c(0, 10, 20, 30, 40),
    colors = c("#313695", "#4575B4", "#74ADD1", "#FEE090", "#F46D43", "#A50026"),
    type = "continuous",
    position = "bottom-left",
    add = TRUE
  )

# Display the map
map
```

![image](https://github.com/user-attachments/assets/b1a16570-bdda-4290-bd89-085dbc4aaf4f)

**Comment by CIOData** [comment]
Thanks. That works, but necessitates one additional feature add—the ability to clear a specific legend by its id in `clear_legend()`.

**Comment by walkerke** [comment]
Ah OK.  I had updated `clear_legend()` to clear all legends from the map.  I can take a look at that.

**Comment by walkerke** [comment]
This is now implemented!  Try it out: 

```r
library(shiny)
library(mapgl)
library(dplyr)

ui <- fluidPage(
  titlePanel("mapboxgl Legend Demo"),
  
  sidebarLayout(
    sidebarPanel(
      actionButton("add_categorical", "Add Categorical Legend"),
      actionButton("add_continuous", "Add Continuous Legend"),
      actionButton("clear_categorical", "Clear Categorical Legend"),
      actionButton("clear_continuous", "Clear Continuous Legend"),
      actionButton("clear_all", "Clear All Legends")
    ),
    
    mainPanel(
      mapboxglOutput("map")
    )
  )
)

server <- function(input, output, session) {
  
  pts <- sf::st_sf(
    geometry = sf::st_sfc(
      sf::st_point(c(-74.5, 40)),
      sf::st_point(c(-73.9, 40.7)),
      sf::st_point(c(-73.8, 40.8)), 
      crs = 4326
    ), 
    size = c(10, 15, 20),
    color = c("red", "green", "blue"))
  
  output$map <- renderMapboxgl({
    mapboxgl() %>%
      add_circle_layer(
        id = "point_layer",
        source = pts,
        circle_radius = get_column("size"),
        circle_color = get_column("color")
      ) %>%
      fit_bounds(pts, animate = TRUE, padding = 50)
  })
  
  map_proxy <- mapboxgl_proxy("map")
  
  observeEvent(input$add_categorical, {
    map_proxy %>%
      add_legend(
        legend_title = "Categories",
        values = c("Low", "Medium", "High"),
        colors = c("red", "green", "blue"),
        type = "categorical",
        circular_patches = TRUE,
        sizes = c(10, 15, 20),
        position = "top-left",
        unique_id = "legend-categorical",
        add = TRUE
      )
  })
  
  observeEvent(input$add_continuous, {
    map_proxy %>%
      add_legend(
        legend_title = "Values",
        values = c(0, 50, 100),
        colors = c("yellow", "orange", "red"),
        type = "continuous",
        position = "top-right",
        unique_id = 'legend-continuous',
        add = TRUE
      )
  })
  
  observeEvent(input$clear_categorical, {
    map_proxy %>% clear_legend(legend_ids = "legend-categorical")
  })
  
  observeEvent(input$clear_continuous, {
    map_proxy %>% clear_legend(legend_ids = "legend-continuous")
  })
  
  observeEvent(input$clear_all, {
    map_proxy %>% clear_legend()
  })
}

shinyApp(ui, server)
```

**Comment by CIOData** [comment]
Thank you. Yes, this works. I did run into one additional issue, but I was able to work around it. The situation I am implementing this in is a dynamic Shiny mapping application where multiple legends come, go, and change. In places, it makes sense for me to have a `clear_legend(legend_ids = 'ID')` on a specific ID, but that ID might not always exist when the `observeEvent` is triggered. If a scenario occurs where the `clear_legend(legend_ids = 'ID')` is called and that ID isn't present on the map currently, an error gets thrown. I overcame it with a `tryCatch()`, but it might be nice just to have that built into the function.

For additional context, [here](https://cancerinfocus.uky.edu/appalachia/) is a working example of one application that I've been converting from `leaflet` into `mapgl`. There are still a few things to work through, but `mapgl` is such an incredible upgrade on speed that it's definitely the direction I want to go. Thank you for putting this package together.
